### PR TITLE
8340454: C2 EA asserts with "previous reducible Phi is no longer reducible before SUT"

### PR DESCRIPTION
**Pull request by JohnTortugo**
Please, consider this patch to fix an issue that happens when a Phi previously considered reducible become later irreducible. The overall situation that causes the problem is like so:

- Consider that there are at least 2 scalar replaceable objects (Obj1 and Obj2; Obj2 is stored in a field of Obj1) when we start iterating the loop at escape.cpp:301

- In the first iteration of the loop the call chain starting with `adjust_scalar_replaceable_state` ends up calling `can_reduce_phi` and considering Phi1 as reducible. This Phi has only Obj1 as *SR* input.

- In another iteration of the loop Obj2 is flagged as NSR. For instance, because we are storing Obj2 in an unknown position of an array. This will cause `found_nsr_alloc` to be set to `true`.

After the loop finishes, the execution will go to `find_scalar_replaceable_allocs`. The code will process Obj1, because it's still scalar replaceable, but will find that this object is stored in a field of a - **now** - NSR object. Therefore, correctly, Obj1 will also be marked as NSR. When Obj1 is marked as NSR Phi1 becomes irreducible because it doesn't have any more scalar replaceable input.

The solution I'm proposing is simply revisit the "reducibility" of the Phis when an object is marked as NSR.

---------

### Tests

Win, Mac & Linux tier1-4 on x64 & Aarch64.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340454](https://bugs.openjdk.org/browse/JDK-8340454): C2 EA asserts with "previous reducible Phi is no longer reducible before SUT" (**Bug** - P3)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) Review applies to [0361f272](https://git.openjdk.org/jdk/pull/21778/files/0361f272096c06a036b7c9e24bdfbbf1a7ebd1f5)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21778/head:pull/21778` \
`$ git checkout pull/21778`

Update a local copy of the PR: \
`$ git checkout pull/21778` \
`$ git pull https://git.openjdk.org/jdk.git pull/21778/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21778`

View PR using the GUI difftool: \
`$ git pr show -t 21778`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21778.diff">https://git.openjdk.org/jdk/pull/21778.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21778#issuecomment-2447476918)
</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Seems reasonable.

File: test/hotspot/jtreg/compiler/escapeAnalysis/TestReduceAllocationAndNonReduciblePhi.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ */
+
+/*
+ * @test
+ * @bug 8340454
+ * @summary Check that Reduce Allocation Merges doesn't crash when
+ *          a reducible Phi becomes irreducible after the last of
+ *          its SR inputs is flagged as NSR.
+ * @run main/othervm compiler.escapeAnalysis.TestReduceAllocationAndNonReduciblePhi
+ * @run main compiler.escapeAnalysis.TestReduceAllocationAndNonReduciblePhi
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/escapeAnalysis/TestReduceAllocationAndNonReduciblePhi.java:31]: Why do you need both? Should `@run main/othervm` add `-Xbatch`?

File: test/hotspot/jtreg/compiler/escapeAnalysis/TestReduceAllocationAndNonReduciblePhi.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ */
+
+package compiler.escapeAnalysis;
+
+public class TestReduceAllocationAndNonReduciblePhi {
+    public static void main(String args[]) {
+        int result = 0;
+
+        for (int i=0; i<20000; i++) {
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/escapeAnalysis/TestReduceAllocationAndNonReduciblePhi.java:40]:
```suggestion
        for (int i = 0; i < 20000; i++) {
```

File: test/hotspot/jtreg/compiler/escapeAnalysis/TestReduceAllocationAndNonReduciblePhi.java (hunk 1 above)
Comment (JohnTortugo) [reply test/hotspot/jtreg/compiler/escapeAnalysis/TestReduceAllocationAndNonReduciblePhi.java:31]: In my experiments the issue was reproducing without any flag.

Following your advice I included -Xbatch and the usual CompileCommands as well. I also kept the second "@run" command to run the method also in a no-custom-flags mode.

**Review by TobiHartmann — APPROVED** [review]

**Comment by TobiHartmann** [comment]
All green.

**Comment by JohnTortugo** [comment]
Thank you @TobiHartmann

**Pull request closed without merging**